File: python/src/fibonacci/sse.py
"""Server-Sent Events parser for streaming HTTP responses.

There are two ways to get this wrong that only show up in production, and both
are avoided here deliberately:

1. **Decoding per chunk.** ``chunk.decode("utf-8")`` on each network chunk
   corrupts any multi-byte character that straddles a chunk boundary — em
   dashes, curly quotes, emoji, and every non-Latin script. A model streaming
   token-by-token hits this constantly. We drive a single incremental decoder
   across the whole stream instead.

2. **Splitting per chunk.** Chunk boundaries have nothing to do with line or
   event boundaries. A single ``data:`` payload routinely arrives in three
   pieces, and ``\\r\\n`` can straddle a boundary such that a naive splitter
   sees two line terminators where there is one. We buffer across chunks and
   hold back a trailing lone ``\\r`` until we know what follows it.

The parser follows the WHATWG SSE line grammar (``\\n``, ``\\r\\n``, and bare
``\\r`` all terminate a line; ``:`` prefixes a comment; one optional space is
stripped after the field colon), with one documented deviation noted on
:func:`iter_sse`.
"""

from __future__ import annotations

import codecs
import json
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, List, Optional, Tuple

__all__ = ["SSEEvent", "iter_sse"]

_BOM = "﻿"


@dataclass(frozen=True)
class SSEEvent:
    """One dispatched Server-Sent Event.

    :param event: The ``event:`` field, or ``None`` when the stream did not
        name one (consumers should then fall back to a ``type`` key inside the
        JSON payload, which is what the Responses API does).
    :param data: Concatenated ``data:`` lines, joined with newlines and with no
        trailing newline.
    """

    data: str
    event: Optional[str] = None

    def json(self) -> Any:
        """Parse :attr:`data` as JSON.

        Raises :class:`json.JSONDecodeError` for non-JSON payloads. Sentinel
        frames such as ``[DONE]`` are not JSON, so check for those before
        calling this.
        """
        return json.loads(self.data)


@dataclass
class _EventBuffer:
    """Accumulator for the fields of the event currently being assembled."""

    name: Optional[str] = None
    data: List[str] = field(default_factory=list)

    def reset(self) -> None:
        self.name = None
        self.data = []


async def iter_sse(chunks: AsyncIterator[bytes]) -> AsyncIterator[SSEEvent]:
    """Parse an async iterator of raw bytes into :class:`SSEEvent` objects.

    Deviation from the strict spec: if the stream ends with a pending event
    that was never terminated by a blank line, that event is still dispatched.
    Real servers — including OpenAI-compatible ones behind proxies — drop the
    final blank line when closing the connection, and silently discarding the
    last event would lose the terminal ``response.completed`` frame carrying
    token usage.

    :param chunks: Byte chunks, e.g. ``httpx.Response.aiter_bytes()``.
    """
    decoder = codecs.getincrementaldecoder("utf-8")("replace")
    buffer = ""
    pending = _EventBuffer()
    at_stream_start = True

    async for chunk in chunks:
        text = decoder.decode(chunk)
        if not text:
            continue
        if at_stream_start:
            # A UTF-8 BOM is legal at the head of an event stream and must not
            # end up prefixed to the first field name.
            text = text.lstrip(_BOM)
            at_stream_start = False
        buffer += text

        lines, buffer = _take_lines(buffer, final=False)
        for line in lines:
            event = _feed(pending, line)
            if event is not None:
                yield event

    # Flush whatever the decoder was holding (a truncated multi-byte sequence
    # becomes a replacement character rather than vanishing), then drain.
    buffer += decoder.decode(b"", True)
    lines, remainder = _take_lines(buffer, final=True)
    for line in lines:
        event = _feed(pending, line)
        if event is not None:
            yield event
    if remainder:
        event = _feed(pending, remainder)
        if event is not None:
            yield event

    if pending.data:
        yield SSEEvent(data="\n".join(pending.data), event=pending.name)


def _take_lines(buffer: str, *, final: bool) -> Tuple[List[str], str]:
    """Split complete lines off the front of ``buffer``.

    Returns the complete lines and the unconsumed remainder. A trailing lone
    ``\\r`` is held back unless ``final`` is set, because the very next chunk
    may begin with ``\\n`` and the pair is a single terminator.
    """
    lines: List[str] = []
    start = 0
    index = 0
    length = len(buffer)

    while index < length:
        char = buffer[index]
        if char == "\n":
            lines.append(buffer[start:index])
            index += 1
            start = index
        elif char == "\r":
            if index + 1 == length and not final:
                break
            lines.append(buffer[start:index])
            index += 1
            if index < length and buffer[index] == "\n":
                index += 1
            start = index
        else:
            index += 1

    return lines, buffer[start:]


def _feed(pending: _EventBuffer, line: str) -> Optional[SSEEvent]:
    """Apply one line to the pending event; return an event when one dispatches."""
    if line == "":
        # Blank line dispatches. Per spec an event with no data is discarded
        # rather than dispatched, but the buffers still reset.
        if not pending.data:
            pending.reset()
            return None
        event = SSEEvent(data="\n".join(pending.data), event=pending.name)
        pending.reset()
        return event

    if line.startswith(":"):
        # Comment. Widely used as a keep-alive heartbeat; carries no payload.
        return None

    name, _, value = line.partition(":")
    if value.startswith(" "):
        value = value[1:]

    if name == "event":
        pending.name = value
    elif name == "data":
        pending.data.append(value)
    # `id` and `retry` are parsed and ignored on purpose: this SDK does not
    # reconnect a dropped stream (a resumed completion would interleave
    # duplicate tokens), so last-event-id and retry hints have no consumer.
    return None

File: src/net/sse.ts
/**
 * Server-Sent Events parser.
 *
 * The bug this file exists to avoid: a naive implementation does
 * `chunk.toString().split('\n')` per chunk. TCP does not respect event
 * boundaries, so a single JSON payload routinely arrives split across two
 * chunks — sometimes mid-multi-byte-character. That produces intermittent
 * "Unexpected end of JSON input" that only shows up on slow networks or long
 * completions, which is the worst possible failure mode to debug.
 *
 * So: one persistent buffer, a streaming TextDecoder, and events emitted only
 * on a complete blank-line terminator.
 *
 * Implements the field parsing of the WHATWG event-stream spec that matters
 * here (data, event, id, retry, comments); ignores the reconnection state
 * machine, which is the caller's business.
 */

export interface SseEvent {
  /** The `event:` field, or 'message' when absent, per spec. */
  event: string;
  /** All `data:` lines joined with newline, with the single leading space stripped. */
  data: string;
  id?: string;
  retry?: number;
}

/** Parse one already-complete event block (no blank lines inside). */
export function parseEventBlock(block: string): SseEvent | null {
  let event = 'message';
  const dataLines: string[] = [];
  let id: string | undefined;
  let retry: number | undefined;
  let sawField = false;

  for (const rawLine of block.split('\n')) {
    const line = rawLine.endsWith('\r') ? rawLine.slice(0, -1) : rawLine;
    if (line === '') continue;
    // A line starting with ':' is a comment. Servers send these as keepalives.
    if (line.startsWith(':')) continue;

    const colon = line.indexOf(':');
    const field = colon === -1 ? line : line.slice(0, colon);
    // Spec: strip exactly one leading space from the value, not all whitespace.
    let value = colon === -1 ? '' : line.slice(colon + 1);
    if (value.startsWith(' ')) value = value.slice(1);

    switch (field) {
      case 'data':
        dataLines.push(value);
        sawField = true;
        break;
      case 'event':
        event = value;
        sawField = true;
        break;
      case 'id':
        // Spec: ignore an id containing NUL.
        if (!value.includes('\0')) id = value;
        sawField = true;
        break;
      case 'retry': {
        const n = Number.parseInt(value, 10);
        if (Number.isFinite(n) && /^\d+$/.test(value)) retry = n;
        sawField = true;
        break;
      }
      default:
        // Unknown fields are ignored per spec.
        break;
    }
  }

  if (!sawField) return null;
  const out: SseEvent = { event, data: dataLines.join('\n') };
  if (id !== undefined) out.id = id;
  if (retry !== undefined) out.retry = retry;
  return out;
}

/**
 * Turn a byte stream into SSE events.
 *
 * `stream: true` on the decoder is what makes multi-byte characters split
 * across chunk boundaries decode correctly instead of becoming U+FFFD.
 */
export async function* parseSse(
  body: ReadableStream<Uint8Array>,
  signal?: AbortSignal,
): AsyncGenerator<SseEvent> {
  const reader = body.getReader();
  const decoder = new TextDecoder('utf-8');
  let buffer = '';

  const onAbort = () => void reader.cancel().catch(() => {});
  signal?.addEventListener('abort', onAbort, { once: true });

  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      // Events are separated by a blank line. Handle \n\n and \r\n\r\n.
      for (;;) {
        const match = /\r?\n\r?\n/.exec(buffer);
        if (!match) break;
        const block = buffer.slice(0, match.index);
        buffer = buffer.slice(match.index + match[0].length);
        const evt = parseEventBlock(block);
        if (evt) yield evt;
      }
    }

    // Flush any trailing bytes and emit a final unterminated block. Some
    // servers close without the final blank line.
    buffer += decoder.decode();
    if (buffer.trim() !== '') {
      const evt = parseEventBlock(buffer);
      if (evt) yield evt;
    }
  } finally {
    signal?.removeEventListener('abort', onAbort);
    reader.releaseLock();
  }
}

/**
 * Convenience wrapper: yield JSON-decoded `data` payloads, skipping the
 * `[DONE]` sentinel and any payload that is not valid JSON (which servers do
 * emit — Ollama and some proxies send bare keepalive text).
 */
export async function* parseSseJson<T = unknown>(
  body: ReadableStream<Uint8Array>,
  signal?: AbortSignal,
): AsyncGenerator<{ event: string; json: T }> {
  for await (const evt of parseSse(body, signal)) {
    const data = evt.data.trim();
    if (data === '' || data === '[DONE]') continue;
    let json: T;
    try {
      json = JSON.parse(data) as T;
    } catch {
      continue;
    }
    yield { event: evt.event, json };
  }
}
